cmd/gcs/main.go

```diff
@@ -81,6 +87,104 @@ func readMemoryEvents(startTime time.Time, efdFile *os.File, cgName string, thre
 	}
 }
 
+// runWithRestartMonitor starts a command with given args and waits for it to exit. If the
+// command exit code is non-zero the command is restarted with with some back off delay.
+// Any stdout or stderr of the command will be split into lines and written as a log with
+// logrus standard logger.  This function must be called in a separate goroutine.
+func runWithRestartMonitor(arg0 string, args ...string) {
+	backoffSettings := backoff.NewExponentialBackOff()
+	// give after retrying for 10 mins
+	backoffSettings.MaxElapsedTime = time.Minute * 10
+	for {
+		command := exec.Command(arg0, args...)
+		err := command.Start()
+		if err != nil {
+			logrus.WithFields(logrus.Fields{
+				"error":   err,
+				"command": command.Args,
+			}).Warn("restart monitor: start returns error")
+		} else {
+			waitErr := command.Wait()
+			if waitErr != nil {
+				logrus.WithFields(logrus.Fields{
+					"error":   waitErr,
+					"command": command.Args,
+				}).Warn("restart monitor: wait returns error")
+			}
+		}
+		backOffTime := backoffSettings.NextBackOff()
+		if backOffTime == backoffSettings.Stop {
+			logrus.WithFields(logrus.Fields{
+				"command": command.Args,
+			}).Warn("give up restarting after timeout")
+			return
+		}
+		time.Sleep(backOffTime)
+	}
+
+}
+
+// StartTimeSyncService starts the `chronyd` deamon to keep the UVM time synchronized.  We
+// use a PTP device provided by the hypervisor as a source of correct time (instead of
+// using a network server). We need to create a configuration file that configures chronyd
+// to use the PTP device.  The system can have multiple PTP devices so we identify the
+// correct PTP device by verifying that the `clock_name` of that device is `hyperv`.
+func StartTimeSyncService() error {
+	ptpClassDir, err := os.Open("/sys/class/ptp")
+	if err != nil {
+		return errors.Wrap(err, "failed to open PTP class directory")
+	}
+
+	ptpDirList, err := ptpClassDir.Readdirnames(-1)
+	if err != nil {
+		return errors.Wrap(err, "failed to list PTP class directory")
+	}
+
+	var ptpDirPath string
+	found := false
+	expectedClockName := "hyperv"
+	for _, ptpDirPath = range ptpDirList {
+		clockNameFilePath := filepath.Join(ptpClassDir.Name(), ptpDirPath, "clock_name")
+		clockNameFile, err := os.Open(clockNameFilePath)
+		if err != nil {
+			return errors.Wrapf(err, "failed to open clock name file at %s", clockNameFilePath)
+		}
+		// Expected clock name is `hyperv` so read first 6 chars and verify the
+		// name
+		buf := make([]byte, len(expectedClockName))
+		fileReader := bufio.NewReader(clockNameFile)
+		_, err = fileReader.Read(buf)
```

Comment:
`Read` (even `(*bufio.Reader).Read`) is not guaranteed to fill the full buffer. We should use `io.ReadFull` instead if we want to ensure we read the whole thing (or switch to `ioutil.ReadFile` as suggested above if applicable).

While this is probably never going to be a problem for reading 6 bytes, there's still a risk that this code could be modified or copied in the future, exposing this bug.

Reply:
Good catch. Fixed.